cpp/src/io/parquet/rle_stream.cuh

```diff
@@ -313,54 +257,97 @@ struct rle_stream {
-                                    min(run.remaining, (output_count - run.output_pos)));
-        batch.decode(end, level_bits, warp_lane, warp_decode_id);
-        // last warp updates total values processed
-        if (warp_lane == 0 && warp_decode_id == num_runs - 1) {
-          values_processed = run.output_pos + batch.size;
+        if (warp_lane == 0) {
+          fill_run_batch();
+          if (decode_index == -1) {
+            // first time, set it to the beginning of the buffer (rolled)
+            decode_index        = 0;
+            decode_index_shared = decode_index;
+          }
+          fill_index_shared = fill_index;
         }
       }
-      __syncthreads();
-
-      // if we haven't run out of space, retrieve the next batch. otherwise leave it for the next
-      // call.
-      if (!t && values_processed < output_count) {
-        thrust::tie(run_start, num_runs) = get_run_batch();
+      // remaining warps decode the runs, starting on the second iteration of this
+      // do/while loop (decode_index == -1 means "first iteration", so we should skip decoding)
+      else if (decode_index >= 0 && decode_index + warp_decode_id < fill_index) {
+        int const run_index = decode_index + warp_decode_id;
+        auto& run           = runs[rolling_index<run_buffer_size>(run_index)];
+        // this is the total amount (absolute) we will write in this invocation
+        // of `decode_next`.
+        int const max_count = cur_values + output_count;
+        // run.output_pos is absolute position, we start decoding
+        // if it's supposed to fit in this call to `decode_next`.
+        if (max_count > run.output_pos) {
+          int remaining        = run.remaining;
+          int const run_offset = run.size - remaining;
+          // last_run_pos is the absolute position of the run, including
+          // what was decoded last time.
+          int const last_run_pos = run.output_pos + run_offset;
+
+          // the amount we should process is the smallest of current remaining, or
+          // space available in the output buffer (for that last run at the end of
+          // a call to decode_next).
+          int const batch_len = min(remaining, max_count - last_run_pos);
+          decode<level_t, max_output_values>(output,
+                                             run.start,
+                                             run_offset,
+                                             run.output_pos,
+                                             run.level_run,
+                                             batch_len,
+                                             end,
+                                             level_bits,
+                                             warp_lane);
+          if (warp_lane == 0) {
+            // after writing this batch, are we at the end of the output buffer?
+            auto const at_end = (last_run_pos + batch_len - cur_values == output_count);
```

Comment:
Parenthesize the arithmetic here :)

Reply:
done https://github.com/rapidsai/cudf/pull/15159/commits/0470a7d32ab9fe5b49bcb3d5d47d6985e15098c9